### ISSUE TITLE
Out of sync applications are referenced in PRs that did not cause the change

### ISSUE THREAD

**Issue by rouke-broersma** [infro-io/issues#2]
I had merged a change to `cilium` on which I do not have autosync enabled. The pending change on cilium shows up in all other open pull requests. This suggests that infro runs the diff on the complete argocd instance instead of only on the changed applications. This seems unnecessary computationally and pollutes the PR diff with unrelated changes.

Example:

https://github.com/broersma-forslund/homelab/pull/152#issuecomment-1952352263

**Comment by colethienes** [comment]
Yeah this is one of the issues with diffing a revision against a live cluster.

>This suggests that infro runs the diff on the complete argocd instance

[This is how](https://infro.io/docs/argo-cd-pr-checks) the PR checks occur.

>instead of only on the changed applications

Infro runs `argo app diff` on all apps with the relevant repo url. Is there a better way to detect changed apps?

**Comment by rouke-broersma** [comment]
> Yeah this is one of the issues with diffing a revision against a live cluster.
> 
> >This suggests that infro runs the diff on the complete argocd instance
> 
> [This is how](https://infro.io/docs/argo-cd-pr-checks) the PR checks occur.
> 
> >instead of only on the changed applications
> 
> Infro runs `argo app diff` on all apps with the relevant repo url. Is there a better way to detect changed apps?

Yes, but not easily. The simplest check would be to check if the commit references a file within the repo path of the application. 

Of course you could be using files outside of that path if you're using kustomize for example. Or if you have a helm dependency on a chart in the same repo. 
To detect those changes you would need to follow references in dependencies and in kustomize overlays. That's more tricky.
There's also argocd annotations you can place to give hints to argocd about your app paths, as an intermediate you could add an option so users can configure infro to only use repo path + app paths. This would make it a conscious choice to enable it and live with potential  accuracy issues if the paths are misconfigured.

**Comment by colethienes** [comment]
I added [functionality](https://github.com/infro-io/infro-core/blob/main/pkg/infro/executor.go#L82) to not mention apps that do not have changes associated! Does that work for now?

**Comment by rouke-broersma** [comment]
It doesn't seem to work, unless I misunderstand what you added. See: https://github.com/broersma-forslund/homelab/pull/207

I had merged a pull request in cilium yesterday however auto-sync is turned off and I have not manually sync'd cilium yet. Now I have a new pull request for democratic-csi. The diff mentions cilium changes as well.

Interestingly enough it shows cilium as having 0 files changed even though the diff shows 3 changed files.
